# series data structure

In [1]:
#pip3 install python
import pandas as pd

In [2]:
animals = ['Tiger', 'Bear', 'Moose']
pd.Series(animals)

0    Tiger
1     Bear
2    Moose
dtype: object

In [3]:
numbers = [1, 2, 3]
pd.Series(numbers)

0    1
1    2
2    3
dtype: int64

In [4]:
best_cricketer = {'Shakib': 'Bangladesh',
          'Kohli': 'India',
          'Smith': 'Australia',
          'Archer': 'England'}
s = pd.Series(best_cricketer)
s

Shakib    Bangladesh
Kohli          India
Smith      Australia
Archer       England
dtype: object

In [5]:
s.index

Index(['Shakib', 'Kohli', 'Smith', 'Archer'], dtype='object')

In [6]:
t = pd.Series(['Tiger', 'Bear', 'Moose'], index=['India', 'America', 'Canada'])
t

India      Tiger
America     Bear
Canada     Moose
dtype: object

In [7]:
import pandas as pd
sports = {'Archery': 'Bhutan',
          'Golf': 'Scotland',
          'Sumo': 'Japan',
          'Taekwondo': 'South Korea'}
s = pd.Series(sports, index=['Golf', 'Sumo', 'Hockey'])
s

Golf      Scotland
Sumo         Japan
Hockey         NaN
dtype: object

In [8]:
# Querying a Series
sports = {'Archery': 'Bhutan',
          'Golf': 'Scotland',
          'Sumo': 'Japan',
          'Taekwondo': 'South Korea'}
s = pd.Series(sports)
#s.iloc[3]
#s.loc['Sumo']
s['Golf']

'Scotland'

In [9]:
s = pd.Series([100.00, 120.00, 101.00, 3.00])
total = 0
for item in s:
    total+=item
print(total)

324.0


In [10]:
# try with numpy
import numpy as np

total = np.sum(s)
print(total)

324.0


In [11]:
#this creates a big series of random numbers
s = pd.Series(np.random.randint(0,1000,10000))
s.head() # will give us some datas

0    960
1    846
2    677
3    419
4    910
dtype: int32

In [12]:
%%timeit -n 1000
summary = 0
for item in s:
    summary+=item

1.32 ms ± 67.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [13]:
%%timeit -n 1000
summary = np.sum(s)

79 µs ± 7.16 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [14]:
original_sports = pd.Series({'Archery': 'Bhutan',
                             'Golf': 'Scotland',
                             'Sumo': 'Japan',
                             'Taekwondo': 'South Korea'})
cricket_loving_countries = pd.Series(['Australia',
                                      'Bangladesh',
                                      'Pakistan',
                                      'England'], 
                                   index=['Cricket',
                                          'Cricket',
                                          'Cricket',
                                          'Cricket'])
all_countries = original_sports.append(cricket_loving_countries)

In [15]:
original_sports

Archery           Bhutan
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
dtype: object

In [16]:
cricket_loving_countries

Cricket     Australia
Cricket    Bangladesh
Cricket      Pakistan
Cricket       England
dtype: object

In [17]:
all_countries

Archery           Bhutan
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
Cricket        Australia
Cricket       Bangladesh
Cricket         Pakistan
Cricket          England
dtype: object

In [18]:
all_countries.loc['Cricket']

Cricket     Australia
Cricket    Bangladesh
Cricket      Pakistan
Cricket       England
dtype: object

# DataFrame Data Structure

In [19]:
import pandas as pd
purchase_1 = pd.Series({'Name': 'Fahim',
                        'Item Purchased': 'Axe Bodywash',
                        'Cost': 22.50})
purchase_2 = pd.Series({'Name': 'Mahi',
                        'Item Purchased': 'Shaving tools',
                        'Cost': 32.50})
purchase_3 = pd.Series({'Name': 'Afra',
                        'Item Purchased': 'Chips',
                        'Cost': 15.00})
df = pd.DataFrame([purchase_1, purchase_2, purchase_3], index=['Store 1', 'Store 1', 'Store 2'])
df.head()

Name Item Purchased  Cost
Store 1  Fahim   Axe Bodywash  22.5
Store 1   Mahi  Shaving tools  32.5
Store 2   Afra          Chips  15.0

In [20]:
df['Date'] = ['December 1', 'January 1', 'July 29']
df

Name Item Purchased  Cost        Date
Store 1  Fahim   Axe Bodywash  22.5  December 1
Store 1   Mahi  Shaving tools  32.5   January 1
Store 2   Afra          Chips  15.0     July 29

In [21]:
df.loc['Store 2']

Name                 Afra
Item Purchased      Chips
Cost                   15
Date              July 29
Name: Store 2, dtype: object

In [22]:
df['Feedback'] = ['Positive', None, 'Negative']
df

Name Item Purchased  Cost        Date  Feedback
Store 1  Fahim   Axe Bodywash  22.5  December 1  Positive
Store 1   Mahi  Shaving tools  32.5   January 1      None
Store 2   Afra          Chips  15.0     July 29  Negative

In [23]:
type(df.loc['Store 2'])

pandas.core.series.Series

In [24]:
df.loc['Store 1']

Name Item Purchased  Cost        Date  Feedback
Store 1  Fahim   Axe Bodywash  22.5  December 1  Positive
Store 1   Mahi  Shaving tools  32.5   January 1      None

In [25]:
df.loc['Store 1', 'Cost']

Store 1    22.5
Store 1    32.5
Name: Cost, dtype: float64

In [26]:
df['Item Purchased']

Store 1     Axe Bodywash
Store 1    Shaving tools
Store 2            Chips
Name: Item Purchased, dtype: object

In [27]:
df.loc['Store 1']['Cost']

Store 1    22.5
Store 1    32.5
Name: Cost, dtype: float64

In [28]:
df.loc[:,['Name', 'Cost']]

Name  Cost
Store 1  Fahim  22.5
Store 1   Mahi  32.5
Store 2   Afra  15.0

In [29]:
df.drop('Store 1')

Name Item Purchased  Cost     Date  Feedback
Store 2  Afra          Chips  15.0  July 29  Negative

In [30]:
copy_df = df.copy()
copy_df = copy_df.drop('Store 1')
copy_df

Name Item Purchased  Cost     Date  Feedback
Store 2  Afra          Chips  15.0  July 29  Negative

In [31]:
del copy_df['Name']
copy_df

Item Purchased  Cost     Date  Feedback
Store 2          Chips  15.0  July 29  Negative

In [32]:
df['Location'] = 'Vancouver','Sylhet','Sylhet'
df

Name Item Purchased  Cost        Date  Feedback   Location
Store 1  Fahim   Axe Bodywash  22.5  December 1  Positive  Vancouver
Store 1   Mahi  Shaving tools  32.5   January 1      None     Sylhet
Store 2   Afra          Chips  15.0     July 29  Negative     Sylhet

In [33]:
# suppose they got a discount on the price,then how would you update it
df['Cost'] *= 0.8
df

Name Item Purchased  Cost        Date  Feedback   Location
Store 1  Fahim   Axe Bodywash  18.0  December 1  Positive  Vancouver
Store 1   Mahi  Shaving tools  26.0   January 1      None     Sylhet
Store 2   Afra          Chips  12.0     July 29  Negative     Sylhet

# Dataframe Indexing and Loading

In [34]:
costs = df['Cost']
costs

Store 1    18.0
Store 1    26.0
Store 2    12.0
Name: Cost, dtype: float64

In [35]:
costs += 2
costs

Store 1    20.0
Store 1    28.0
Store 2    14.0
Name: Cost, dtype: float64

In [36]:
df

Name Item Purchased  Cost        Date  Feedback   Location
Store 1  Fahim   Axe Bodywash  20.0  December 1  Positive  Vancouver
Store 1   Mahi  Shaving tools  28.0   January 1      None     Sylhet
Store 2   Afra          Chips  14.0     July 29  Negative     Sylhet

In [37]:
df['Name'][df['Cost']> 20.0]

Store 1    Mahi
Name: Name, dtype: object

## Working on two dataframe

In [38]:
import pandas as pd
df_1 = pd.DataFrame([{'Name':'Abir','Role':'Programming'},
                   {'Name':'Anjoli','Role':'Web Design'}])

df_1= df_1.set_index('Name')
print(df_1)
df_2 = pd.DataFrame([{'Name':'Akib','Job':'Professor'},
                   {'Name':'Afra','Job':'HR executive'}])

df_2= df_2.set_index('Name')
print(df_2)

               Role
Name               
Abir    Programming
Anjoli   Web Design
               Job
Name              
Akib     Professor
Afra  HR executive


To merge two data frame, we have to consider left and right position,<br>
<b>how:</b> This defines what kind of merge to make. It defaults to 'inner', but other possible options include 'outer', 'left', and 'right'.

In [39]:
pd.merge(df_1, df_2, how='outer', left_index=True, right_index=True)

Role           Job
Name                             
Abir    Programming           NaN
Afra            NaN  HR executive
Akib            NaN     Professor
Anjoli   Web Design           NaN

To avoid the nan issue, we can reset our index.

In [40]:
df_1 = df_1.reset_index()
df_2 = df_2.reset_index()
pd.merge(df_1, df_2, how='outer', left_index=True, right_index=True)

Name_x         Role Name_y           Job
0    Abir  Programming   Akib     Professor
1  Anjoli   Web Design   Afra  HR executive

In [41]:
pd.merge(df_1, df_2, how='left', left_on='Name', right_on='Name')

Name         Role  Job
0    Abir  Programming  NaN
1  Anjoli   Web Design  NaN

In [42]:
staff_df = pd.DataFrame([{'Name': 'Kelly', 'Role': 'Director of HR', 'Location': 'State Street'},
                         {'Name': 'Sally', 'Role': 'Course liasion', 'Location': 'Washington Avenue'},
                         {'Name': 'James', 'Role': 'Grader', 'Location': 'Washington Avenue'}])
student_df = pd.DataFrame([{'Name': 'James', 'School': 'Business', 'Location': '1024 Billiard Avenue'},
                           {'Name': 'Mike', 'School': 'Law', 'Location': 'Fraternity House #22'},
                           {'Name': 'Sally', 'School': 'Engineering', 'Location': '512 Wilson Crescent'}])
pd.merge(staff_df, student_df, how='left', left_on='Name', right_on='Name')

Name            Role         Location_x       School            Location_y
0  Kelly  Director of HR       State Street          NaN                   NaN
1  Sally  Course liasion  Washington Avenue  Engineering   512 Wilson Crescent
2  James          Grader  Washington Avenue     Business  1024 Billiard Avenue

In [43]:
staff_df = pd.DataFrame([{'First Name': 'Kelly', 'Last Name': 'Desjardins', 'Role': 'Director of HR'},
                         {'First Name': 'Sally', 'Last Name': 'Brooks', 'Role': 'Course liasion'},
                         {'First Name': 'James', 'Last Name': 'Wilde', 'Role': 'Grader'}])
student_df = pd.DataFrame([{'First Name': 'James', 'Last Name': 'Hammond', 'School': 'Business'},
                           {'First Name': 'Mike', 'Last Name': 'Smith', 'School': 'Law'},
                           {'First Name': 'Sally', 'Last Name': 'Brooks', 'School': 'Engineering'}])
pd.merge(staff_df, student_df, how='inner', left_on=['First Name','Last Name'], right_on=['First Name','Last Name'])

First Name Last Name            Role       School
0      Sally    Brooks  Course liasion  Engineering

#  Working with CSV

In [44]:
df = pd.read_csv('olympics.csv') # read the file
df.head(5)

0         1     2     3     4      5         6     7     8  \
0                NaN  № Summer  01 !  02 !  03 !  Total  № Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  № Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

In [45]:
# since the first column does not have meaningful datas, so we can remove it
df = pd.read_csv('olympics.csv', index_col = 0, skiprows=1)
df.head()

№ Summer  01 !  02 !  03 !  Total  № Winter  01 !.1  \
Afghanistan (AFG)              13     0     0     2      2         0       0   
Algeria (ALG)                  12     5     2     8     15         3       0   
Argentina (ARG)                23    18    24    28     70        18       0   
Armenia (ARM)                   5     1     2     9     12         6       0   
Australasia (ANZ) [ANZ]         2     3     4     5     12         0       0   

                         02 !.1  03 !.1  Total.1  № Games  01 !.2  02 !.2  \
Afghanistan (AFG)             0       0        0       13       0       0   
Algeria (ALG)                 0       0        0       15       5       2   
Argentina (ARG)               0       0        0       41      18      24   
Armenia (ARM)                 0       0        0       11       1       2   
Australasia (ANZ) [ANZ]       0       0        0        2       3       4   

                         03 !.2  Combined total  
Afghanistan (AFG)             2               2  
Algeria (ALG)                 8              15  
Argentina (ARG)              28              70  
Armenia (ARM)                 9              12  
Australasia (ANZ) [ANZ]       5              12

In [46]:
df.columns

Index(['№ Summer', '01 !', '02 !', '03 !', 'Total', '№ Winter', '01 !.1',
       '02 !.1', '03 !.1', 'Total.1', '№ Games', '01 !.2', '02 !.2', '03 !.2',
       'Combined total'],
      dtype='object')

In [47]:
for col in df.columns:
    if col[:2]=='01':
        df.rename(columns={col:'Gold' + col[4:]}, inplace=True)
    if col[:2]=='02':
        df.rename(columns={col:'Silver' + col[4:]}, inplace=True)
    if col[:2]=='03':
        df.rename(columns={col:'Bronze' + col[4:]}, inplace=True)
    if col[:1]=='№':
        df.rename(columns={col:'#' + col[2:]}, inplace=True) 

df.head()

#Summer  Gold  Silver  Bronze  Total  #Winter  \
Afghanistan (AFG)             13     0       0       2      2        0   
Algeria (ALG)                 12     5       2       8     15        3   
Argentina (ARG)               23    18      24      28     70       18   
Armenia (ARM)                  5     1       2       9     12        6   
Australasia (ANZ) [ANZ]        2     3       4       5     12        0   

                         Gold.1  Silver.1  Bronze.1  Total.1  #Games  Gold.2  \
Afghanistan (AFG)             0         0         0        0      13       0   
Algeria (ALG)                 0         0         0        0      15       5   
Argentina (ARG)               0         0         0        0      41      18   
Armenia (ARM)                 0         0         0        0      11       1   
Australasia (ANZ) [ANZ]       0         0         0        0       2       3   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)               0         2               2  
Algeria (ALG)                   2         8              15  
Argentina (ARG)                24        28              70  
Armenia (ARM)                   2         9              12  
Australasia (ANZ) [ANZ]         4         5              12

# Querying a DataFrame

In [48]:
import pandas as pd

df = pd.read_csv('olympics.csv', index_col=0, skiprows=1)

for col in df.columns:
    if col[:2]=='01':
        df.rename(columns={col:'Gold'+col[4:]}, inplace=True)
    if col[:2]=='02':
        df.rename(columns={col:'Silver'+col[4:]}, inplace=True)
    if col[:2]=='03':
        df.rename(columns={col:'Bronze'+col[4:]}, inplace=True)
    if col[:1]=='№':
        df.rename(columns={col:'#'+col[1:]}, inplace=True)

names_ids = df.index.str.split('\s\(') # split the index by '('

df.index = names_ids.str[0] # the [0] element is the country name (new index) 
df['ID'] = names_ids.str[1].str[:3] # the [1] element is the abbreviation or ID (take first 3 characters from that)

df = df.drop('Totals')
df.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Afghanistan        13     0       0       2      2         0       0   
Algeria            12     5       2       8     15         3       0   
Argentina          23    18      24      28     70        18       0   
Armenia             5     1       2       9     12         6       0   
Australasia         2     3       4       5     12         0       0   

             Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
Afghanistan         0         0        0       13       0         0         2   
Algeria             0         0        0       15       5         2         8   
Argentina           0         0        0       41      18        24        28   
Armenia             0         0        0       11       1         2         9   
Australasia         0         0        0        2       3         4         5   

             Combined total   ID  
Afghanistan               2  AFG  
Algeria                  15  ALG  
Argentina                70  ARG  
Armenia                  12  ARM  
Australasia              12  ANZ

In [49]:
# get idea about a specific country
def countr_y():
    return df.iloc[0]
countr_y()

# Summer           13
Gold                0
Silver              0
Bronze              2
Total               2
# Winter            0
Gold.1              0
Silver.1            0
Bronze.1            0
Total.1             0
# Games            13
Gold.2              0
Silver.2            0
Bronze.2            2
Combined total      2
ID                AFG
Name: Afghanistan, dtype: object

In [50]:
def maximgld():
    return df['Gold'].idxmax() #argmax()
maximgld()


'United States'

In [51]:
def max_goldlost():
    return (df['Gold'] - df['Gold.1']).idxmax()
max_goldlost()

'United States'

In [52]:
# to find the countries who has won only gold in summer
only_gold = df.where(df['Gold'] > 0)
only_gold.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Afghanistan       NaN   NaN     NaN     NaN    NaN       NaN     NaN   
Algeria          12.0   5.0     2.0     8.0   15.0       3.0     0.0   
Argentina        23.0  18.0    24.0    28.0   70.0      18.0     0.0   
Armenia           5.0   1.0     2.0     9.0   12.0       6.0     0.0   
Australasia       2.0   3.0     4.0     5.0   12.0       0.0     0.0   

             Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
Afghanistan       NaN       NaN      NaN      NaN     NaN       NaN       NaN   
Algeria           0.0       0.0      0.0     15.0     5.0       2.0       8.0   
Argentina         0.0       0.0      0.0     41.0    18.0      24.0      28.0   
Armenia           0.0       0.0      0.0     11.0     1.0       2.0       9.0   
Australasia       0.0       0.0      0.0      2.0     3.0       4.0       5.0   

             Combined total   ID  
Afghanistan             NaN  NaN  
Algeria                15.0  ALG  
Argentina              70.0  ARG  
Armenia                12.0  ARM  
Australasia            12.0  ANZ

Which country has the biggest difference between their summer gold medal counts and winter gold medal counts relative to their total gold medal count? 

$$\frac{Summer~Gold - Winter~Gold}{Total~Gold}$$

Only include countries that have won at least 1 gold in both summer and winter.

*This function should return a single string value.*

In [53]:
def aree():
    only_gold = df[(df['Gold.1'] >= 1) & (df['Gold'] >=1)]
    only_gold['Difference'] = (abs(only_gold['Gold'] - only_gold['Gold.1']))/only_gold['Gold.2']
    return only_gold['Difference'].idxmax()
aree()

C:\Users\aBr\anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


'Bulgaria'

In [54]:
# total golds in summer
only_gold['Gold'].count()

99

In [55]:
# total golds in summer+winter
df['Gold'].count()

146

In [56]:
# did not win in summer but won in winter
df[(df['Gold.1'] > 0) & (df['Gold'] == 0)]

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Liechtenstein        16     0       0       0      0        18       2   

               Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  \
Liechtenstein         2         5        9       34       2         2   

               Bronze.2  Combined total   ID  
Liechtenstein         5               9  LIE

# Indexing Dataframes

In [57]:
df.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Afghanistan        13     0       0       2      2         0       0   
Algeria            12     5       2       8     15         3       0   
Argentina          23    18      24      28     70        18       0   
Armenia             5     1       2       9     12         6       0   
Australasia         2     3       4       5     12         0       0   

             Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  \
Afghanistan         0         0        0       13       0         0         2   
Algeria             0         0        0       15       5         2         8   
Argentina           0         0        0       41      18        24        28   
Armenia             0         0        0       11       1         2         9   
Australasia         0         0        0        2       3         4         5   

             Combined total   ID  
Afghanistan               2  AFG  
Algeria                  15  ALG  
Argentina                70  ARG  
Armenia                  12  ARM  
Australasia              12  ANZ

# Indexing Dataframes

In [58]:
df['country'] = df.index
df = df.set_index('Gold')
df.head()

# Summer  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  Bronze.1  \
Gold                                                                          
0           13       0       2      2         0       0         0         0   
5           12       2       8     15         3       0         0         0   
18          23      24      28     70        18       0         0         0   
1            5       2       9     12         6       0         0         0   
3            2       4       5     12         0       0         0         0   

      Total.1  # Games  Gold.2  Silver.2  Bronze.2  Combined total   ID  \
Gold                                                                      
0           0       13       0         0         2               2  AFG   
5           0       15       5         2         8              15  ALG   
18          0       41      18        24        28              70  ARG   
1           0       11       1         2         9              12  ARM   
3           0        2       3         4         5              12  ANZ   

          country  
Gold               
0     Afghanistan  
5         Algeria  
18      Argentina  
1         Armenia  
3     Australasia

In [59]:
df = df.reset_index()
df.head()

Gold  # Summer  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  \
0     0        13       0       2      2         0       0         0   
1     5        12       2       8     15         3       0         0   
2    18        23      24      28     70        18       0         0   
3     1         5       2       9     12         6       0         0   
4     3         2       4       5     12         0       0         0   

   Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  Combined total  \
0         0        0       13       0         0         2               2   
1         0        0       15       5         2         8              15   
2         0        0       41      18        24        28              70   
3         0        0       11       1         2         9              12   
4         0        0        2       3         4         5              12   

    ID      country  
0  AFG  Afghanistan  
1  ALG      Algeria  
2  ARG    Argentina  
3  ARM      Armenia  
4  ANZ  Australasia

a function to update the dataframe to include a new column called "Points" which is a weighted value where each gold medal counts for 3 points, silver medals for 2 points, and bronze mdeals for 1 point.

In [60]:
def updated():
    df['Points'] = df['Gold.2']*3 + df['Silver.2']*2 + df['Bronze.2']*1
    return df['Points']
updated()

0        2
1       27
2      130
3       16
4       22
      ... 
141    171
142      4
143      3
144     18
145     38
Name: Points, Length: 146, dtype: int64

# sorting

In [61]:
# Setup
np.random.seed(0)
df = pd.DataFrame({'A': list('accab'), 'B': np.random.choice(10, 5)})    
df

A  B
0  a  5
1  c  0
2  c  3
3  a  3
4  b  7

**Sort by Single Column** <br>
For example, to sort df by column "A", use sort_values with a single column name:

In [62]:
df.sort_values(by='A')

A  B
0  a  5
3  a  3
4  b  7
1  c  0
2  c  3

In [63]:
# Sort by Multiple Columns
df.sort_values(by=['A','B'])

A  B
3  a  3
0  a  5
4  b  7
1  c  0
2  c  3

In [64]:
#Sort By DataFrame Index
df2 = df.sample(frac=1)
df2

A  B
2  c  3
0  a  5
4  b  7
3  a  3
1  c  0

In [65]:
%timeit df2.sort_index() 

280 µs ± 24.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [74]:
census_df = pd.read_csv('census.csv')
census_df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  RDOMESTICMIG2011  \
0        4779736            4780127          4785161  ...          0.002295   
1          54571              54571            54660  ...          7.242091   
2         182265             182265           183193  ...         14.832960   
3          27457              27457            27341  ...         -4.728132   
4          22915              22919            22861  ...         -5.527043   

   RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  RDOMESTICMIG2015  \
0         -0.193196          0.381066          0.582002         -0.467369   
1         -2.915927         -3.012349          2.265971         -2.530799   
2         17.647293         21.845705         19.243287         17.197872   
3         -2.500690         -7.056824         -3.904217        -10.543299   
4         -5.068871         -6.201001         -0.177537          0.177258   

   RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  
0     1.030015     0.826644     1.383282     1.724718     0.712594  
1     7.606016    -2.626146    -2.722002     2.592270    -2.187333  
2    15.844176    18.559627    22.727626    20.317142    18.293499  
3    -4.874741    -2.758113    -7.167664    -3.978583   -10.543299  
4    -5.088389    -4.363636    -5.403729     0.754533     1.107861  

[5 rows x 100 columns]

Which state has the most counties in it

In [75]:
def ans():
    df=census_df[census_df['SUMLEV'] == 50]
    df = df.groupby( [ "SUMLEV", "STNAME"] ).size().to_frame(name = 'count').reset_index()
    return df.loc[df['count'].idxmax()]['STNAME']
ans()  

'Texas'

Only looking at the three most populous counties for each state, what are the three most populous states (in order of highest population to lowest population)? Use CENSUS2010POP.

In [76]:
def mostthree():
    df=census_df[census_df['SUMLEV'] == 50]
    df1=df.sort_values(by=['STNAME','POPESTIMATE2015'],ascending=False).groupby('STNAME').head(3).copy()
    df2 = df1.reset_index().groupby("STNAME").sum().sort_values(by=['POPESTIMATE2015'],ascending=False).head(3).copy()
    return list(df2.index.values)
mostthree()

['California', 'Texas', 'Illinois']

Which county has had the largest absolute change in population within the period 2010-2015? (Hint: population values are stored in columns POPESTIMATE2010 through POPESTIMATE2015, you need to consider all six columns.)

e.g. If County Population in the 5 year period is 100, 120, 80, 105, 100, 130, then its largest change in the period would be |130-80| = 50.

In [77]:
import pandas as pd
pd.options.mode.chained_assignment = None  # default='warn'
census_df = pd.read_csv('census.csv')
def county():
    df=census_df[census_df['SUMLEV'] == 50]
    df['STDEV'] = df[['POPESTIMATE2010','POPESTIMATE2011','POPESTIMATE2012',
                      'POPESTIMATE2013','POPESTIMATE2014','POPESTIMATE2015']].std(axis=1)
    return df.loc[df['STDEV'].idxmax()]['CTYNAME']
county()

'Harris County'

In this datafile, the United States is broken up into four regions using the "REGION" column. 

Create a query that finds the counties that belong to regions 1 or 2, whose name starts with 'Washington', and whose POPESTIMATE2015 was greater than their POPESTIMATE 2014.

*This function should return a 5x2 DataFrame with the columns = ['STNAME', 'CTYNAME'] and the same index ID as the census_df (sorted ascending by index).*

In [78]:
def answer():
    df=census_df[census_df['SUMLEV'] == 50]
    region1 = df['REGION'] == 1
    region2 = df['REGION'] == 2
    county_name = df['CTYNAME'].str.startswith('Washington')
    pop_estim = df['POPESTIMATE2015'] > df['POPESTIMATE2014']
    df = df[(region2|region1)&county_name&pop_estim]
    return df[['STNAME', 'CTYNAME']]
answer()

STNAME            CTYNAME
896           Iowa  Washington County
1419     Minnesota  Washington County
2345  Pennsylvania  Washington County
2355  Rhode Island  Washington County
3163     Wisconsin  Washington County

In [79]:
df=census_df[census_df['SUMLEV'] == 50]
df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   
5      50       3         6      1       9  Alabama   Blount County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  RDOMESTICMIG2011  \
1          54571              54571            54660  ...          7.242091   
2         182265             182265           183193  ...         14.832960   
3          27457              27457            27341  ...         -4.728132   
4          22915              22919            22861  ...         -5.527043   
5          57322              57322            57373  ...          1.807375   

   RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  RDOMESTICMIG2015  \
1         -2.915927         -3.012349          2.265971         -2.530799   
2         17.647293         21.845705         19.243287         17.197872   
3         -2.500690         -7.056824         -3.904217        -10.543299   
4         -5.068871         -6.201001         -0.177537          0.177258   
5         -1.177622         -1.748766         -2.062535         -1.369970   

   RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  
1     7.606016    -2.626146    -2.722002     2.592270    -2.187333  
2    15.844176    18.559627    22.727626    20.317142    18.293499  
3    -4.874741    -2.758113    -7.167664    -3.978583   -10.543299  
4    -5.088389    -4.363636    -5.403729     0.754533     1.107861  
5     1.859511    -0.848580    -1.402476    -1.577232    -0.884411  

[5 rows x 100 columns]

In [80]:
columns_to_keep = ['STNAME',
                   'CTYNAME',
                   'BIRTHS2010',
                   'BIRTHS2011',
                   'BIRTHS2012',
                   'BIRTHS2013',
                   'BIRTHS2014',
                   'BIRTHS2015',
                   'POPESTIMATE2010',
                   'POPESTIMATE2011',
                   'POPESTIMATE2012',
                   'POPESTIMATE2013',
                   'POPESTIMATE2014',
                   'POPESTIMATE2015']
df = df[columns_to_keep]
df.head()

STNAME         CTYNAME  BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
1  Alabama  Autauga County         151         636         615         574   
2  Alabama  Baldwin County         517        2187        2092        2160   
3  Alabama  Barbour County          70         335         300         283   
4  Alabama     Bibb County          44         266         245         259   
5  Alabama   Blount County         183         744         710         646   

   BIRTHS2014  BIRTHS2015  POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  \
1         623         600            54660            55253            55175   
2        2186        2240           183193           186659           190396   
3         260         269            27341            27226            27159   
4         247         253            22861            22733            22642   
5         618         603            57373            57711            57776   

   POPESTIMATE2013  POPESTIMATE2014  POPESTIMATE2015  
1            55038            55290            55347  
2           195126           199713           203709  
3            26973            26815            26489  
4            22512            22549            22583  
5            57734            57658            57673

In [81]:
df = df.set_index(['STNAME', 'CTYNAME'])
df.head()

BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
STNAME  CTYNAME                                                          
Alabama Autauga County         151         636         615         574   
        Baldwin County         517        2187        2092        2160   
        Barbour County          70         335         300         283   
        Bibb County             44         266         245         259   
        Blount County          183         744         710         646   

                        BIRTHS2014  BIRTHS2015  POPESTIMATE2010  \
STNAME  CTYNAME                                                   
Alabama Autauga County         623         600            54660   
        Baldwin County        2186        2240           183193   
        Barbour County         260         269            27341   
        Bibb County            247         253            22861   
        Blount County          618         603            57373   

                        POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME  CTYNAME                                                             
Alabama Autauga County            55253            55175            55038   
        Baldwin County           186659           190396           195126   
        Barbour County            27226            27159            26973   
        Bibb County               22733            22642            22512   
        Blount County             57711            57776            57734   

                        POPESTIMATE2014  POPESTIMATE2015  
STNAME  CTYNAME                                           
Alabama Autauga County            55290            55347  
        Baldwin County           199713           203709  
        Barbour County            26815            26489  
        Bibb County               22549            22583  
        Blount County             57658            57673

# Idiomatic Pandas: Making Code Pandorable


In [82]:
import pandas as pd
df = pd.read_csv('census.csv')
df

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME            CTYNAME  \
0         40       3         6      1       0  Alabama            Alabama   
1         50       3         6      1       1  Alabama     Autauga County   
2         50       3         6      1       3  Alabama     Baldwin County   
3         50       3         6      1       5  Alabama     Barbour County   
4         50       3         6      1       7  Alabama        Bibb County   
...      ...     ...       ...    ...     ...      ...                ...   
3188      50       4         8     56      37  Wyoming  Sweetwater County   
3189      50       4         8     56      39  Wyoming       Teton County   
3190      50       4         8     56      41  Wyoming       Uinta County   
3191      50       4         8     56      43  Wyoming    Washakie County   
3192      50       4         8     56      45  Wyoming      Weston County   

      CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  \
0           4779736            4780127          4785161  ...   
1             54571              54571            54660  ...   
2            182265             182265           183193  ...   
3             27457              27457            27341  ...   
4             22915              22919            22861  ...   
...             ...                ...              ...  ...   
3188          43806              43806            43593  ...   
3189          21294              21294            21297  ...   
3190          21118              21118            21102  ...   
3191           8533               8533             8545  ...   
3192           7208               7208             7181  ...   

      RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
0             0.002295         -0.193196          0.381066          0.582002   
1             7.242091         -2.915927         -3.012349          2.265971   
2            14.832960         17.647293         21.845705         19.243287   
3            -4.728132         -2.500690         -7.056824         -3.904217   
4            -5.527043         -5.068871         -6.201001         -0.177537   
...                ...               ...               ...               ...   
3188          1.072643         16.243199         -5.339774        -14.252889   
3189         -1.589565          0.972695         19.525929         14.143021   
3190        -17.755986         -4.916350         -6.902954        -14.215862   
3191        -11.637475         -0.827815         -2.013502        -17.781491   
3192        -11.752361         -8.040059         12.372583          1.533635   

      RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
0            -0.467369     1.030015     0.826644     1.383282     1.724718   
1            -2.530799     7.606016    -2.626146    -2.722002     2.592270   
2            17.197872    15.844176    18.559627    22.727626    20.317142   
3           -10.543299    -4.874741    -2.758113    -7.167664    -3.978583   
4             0.177258    -5.088389    -4.363636    -5.403729     0.754533   
...                ...          ...          ...          ...          ...   
3188        -14.248864     1.255221    16.243199    -5.295460   -14.075283   
3189         -0.564849     0.654527     2.408578    21.160658    16.308671   
3190        -12.127022   -18.136812    -5.536861    -7.521840   -14.740608   
3191          1.682288   -11.990126    -1.182592    -2.250385   -18.020168   
3192          6.935294   -12.032179    -8.040059    12.372583     1.533635   

      RNETMIG2015  
0        0.712594  
1       -2.187333  
2       18.293499  
3      -10.543299  
4        1.107861  
...           ...  
3188   -14.070195  
3189     1.520747  
3190   -12.606351  
3191     1.441961  
3192     6.935294  

[3193 rows x 100 columns]

In [83]:
(df.where(df['SUMLEV']==50)
    .dropna()
    .set_index(['STNAME','CTYNAME'])
    .rename(columns={'ESTIMATESBASE2010': 'Estimates Base 2010'}))

SUMLEV  REGION  DIVISION  STATE  COUNTY  \
STNAME  CTYNAME                                                      
Alabama Autauga County       50.0     3.0       6.0    1.0     1.0   
        Baldwin County       50.0     3.0       6.0    1.0     3.0   
        Barbour County       50.0     3.0       6.0    1.0     5.0   
        Bibb County          50.0     3.0       6.0    1.0     7.0   
        Blount County        50.0     3.0       6.0    1.0     9.0   
...                           ...     ...       ...    ...     ...   
Wyoming Sweetwater County    50.0     4.0       8.0   56.0    37.0   
        Teton County         50.0     4.0       8.0   56.0    39.0   
        Uinta County         50.0     4.0       8.0   56.0    41.0   
        Washakie County      50.0     4.0       8.0   56.0    43.0   
        Weston County        50.0     4.0       8.0   56.0    45.0   

                           CENSUS2010POP  Estimates Base 2010  \
STNAME  CTYNAME                                                 
Alabama Autauga County           54571.0              54571.0   
        Baldwin County          182265.0             182265.0   
        Barbour County           27457.0              27457.0   
        Bibb County              22915.0              22919.0   
        Blount County            57322.0              57322.0   
...                                  ...                  ...   
Wyoming Sweetwater County        43806.0              43806.0   
        Teton County             21294.0              21294.0   
        Uinta County             21118.0              21118.0   
        Washakie County           8533.0               8533.0   
        Weston County             7208.0               7208.0   

                           POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  \
STNAME  CTYNAME                                                                
Alabama Autauga County             54660.0          55253.0          55175.0   
        Baldwin County            183193.0         186659.0         190396.0   
        Barbour County             27341.0          27226.0          27159.0   
        Bibb County                22861.0          22733.0          22642.0   
        Blount County              57373.0          57711.0          57776.0   
...                                    ...              ...              ...   
Wyoming Sweetwater County          43593.0          44041.0          45104.0   
        Teton County               21297.0          21482.0          21697.0   
        Uinta County               21102.0          20912.0          20989.0   
        Washakie County             8545.0           8469.0           8443.0   
        Weston County               7181.0           7114.0           7065.0   

                           ...  RDOMESTICMIG2011  RDOMESTICMIG2012  \
STNAME  CTYNAME            ...                                       
Alabama Autauga County     ...          7.242091         -2.915927   
        Baldwin County     ...         14.832960         17.647293   
        Barbour County     ...         -4.728132         -2.500690   
        Bibb County        ...         -5.527043         -5.068871   
        Blount County      ...          1.807375         -1.177622   
...                        ...               ...               ...   
Wyoming Sweetwater County  ...          1.072643         16.243199   
        Teton County       ...         -1.589565          0.972695   
        Uinta County       ...        -17.755986         -4.916350   
        Washakie County    ...        -11.637475         -0.827815   
        Weston County      ...        -11.752361         -8.040059   

                           RDOMESTICMIG2013  RDOMESTICMIG2014  \
STNAME  CTYNAME                                                 
Alabama Autauga County            -3.012349          2.265971   
        Baldwin County            21.845705         19.243287   
        Barbour County            -7.056824         -3.904

In [84]:
df = df[df['SUMLEV']==50]
df.set_index(['STNAME','CTYNAME'], inplace=True)
df.rename(columns={'ESTIMATESBASE2010': 'Estimates Base 2010'})

SUMLEV  REGION  DIVISION  STATE  COUNTY  \
STNAME  CTYNAME                                                      
Alabama Autauga County         50       3         6      1       1   
        Baldwin County         50       3         6      1       3   
        Barbour County         50       3         6      1       5   
        Bibb County            50       3         6      1       7   
        Blount County          50       3         6      1       9   
...                           ...     ...       ...    ...     ...   
Wyoming Sweetwater County      50       4         8     56      37   
        Teton County           50       4         8     56      39   
        Uinta County           50       4         8     56      41   
        Washakie County        50       4         8     56      43   
        Weston County          50       4         8     56      45   

                           CENSUS2010POP  Estimates Base 2010  \
STNAME  CTYNAME                                                 
Alabama Autauga County             54571                54571   
        Baldwin County            182265               182265   
        Barbour County             27457                27457   
        Bibb County                22915                22919   
        Blount County              57322                57322   
...                                  ...                  ...   
Wyoming Sweetwater County          43806                43806   
        Teton County               21294                21294   
        Uinta County               21118                21118   
        Washakie County             8533                 8533   
        Weston County               7208                 7208   

                           POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  \
STNAME  CTYNAME                                                                
Alabama Autauga County               54660            55253            55175   
        Baldwin County              183193           186659           190396   
        Barbour County               27341            27226            27159   
        Bibb County                  22861            22733            22642   
        Blount County                57373            57711            57776   
...                                    ...              ...              ...   
Wyoming Sweetwater County            43593            44041            45104   
        Teton County                 21297            21482            21697   
        Uinta County                 21102            20912            20989   
        Washakie County               8545             8469             8443   
        Weston County                 7181             7114             7065   

                           ...  RDOMESTICMIG2011  RDOMESTICMIG2012  \
STNAME  CTYNAME            ...                                       
Alabama Autauga County     ...          7.242091         -2.915927   
        Baldwin County     ...         14.832960         17.647293   
        Barbour County     ...         -4.728132         -2.500690   
        Bibb County        ...         -5.527043         -5.068871   
        Blount County      ...          1.807375         -1.177622   
...                        ...               ...               ...   
Wyoming Sweetwater County  ...          1.072643         16.243199   
        Teton County       ...         -1.589565          0.972695   
        Uinta County       ...        -17.755986         -4.916350   
        Washakie County    ...        -11.637475         -0.827815   
        Weston County      ...        -11.752361         -8.040059   

                           RDOMESTICMIG2013  RDOMESTICMIG2014  \
STNAME  CTYNAME                                                 
Alabama Autauga County            -3.012349          2.265971   
        Baldwin County            21.845705         19.243287   
        Barbour County            -7.056824         -3.904

In [85]:
import numpy as np
def min_max(row):
    data = row[['POPESTIMATE2010',
                'POPESTIMATE2011',
                'POPESTIMATE2012',
                'POPESTIMATE2013',
                'POPESTIMATE2014',
                'POPESTIMATE2015']]
    return pd.Series({'min': np.min(data), 'max': np.max(data)})

In [86]:
df.apply(min_max, axis=1) # get output

min       max
STNAME  CTYNAME                              
Alabama Autauga County      54660.0   55347.0
        Baldwin County     183193.0  203709.0
        Barbour County      26489.0   27341.0
        Bibb County         22512.0   22861.0
        Blount County       57373.0   57776.0
...                             ...       ...
Wyoming Sweetwater County   43593.0   45162.0
        Teton County        21297.0   23125.0
        Uinta County        20822.0   21102.0
        Washakie County      8316.0    8545.0
        Weston County        7065.0    7234.0

[3142 rows x 2 columns]

Or we could have added the result directly to the file by

In [ ]:
import numpy as np
def min_max(row):
    data = row[['POPESTIMATE2010',
                'POPESTIMATE2011',
                'POPESTIMATE2012',
                'POPESTIMATE2013',
                'POPESTIMATE2014',
                'POPESTIMATE2015']]
    row['max'] = np.max(data)
    row['min'] = np.min(data)
    return row
df.apply(min_max, axis=1)

In [ ]:
rows = ['POPESTIMATE2010',
        'POPESTIMATE2011',
        'POPESTIMATE2012',
        'POPESTIMATE2013',
        'POPESTIMATE2014',
        'POPESTIMATE2015']
df.apply(lambda x: np.max(x[rows]), axis=1)

# Missing Data

In [ ]:
import pandas as pd

In [ ]:
df =  pd.read_csv('log.csv')
df.fillna #fill with NA with empty cases

In [ ]:
df = df.set_index('time')
df = df.sort_index()
df

Observe that index has same values and we can improve it

In [ ]:
df = df.reset_index()
df = df.set_index(['user','time'])
df

In [ ]:
df = df.fillna(method='ffill')
df.head()

# Group by

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv('census.csv')
df = df[df['SUMLEV']==50]
df

In [ ]:
df = df.set_index('STNAME')

def fun(item):
    if item[0]<'M':
        return 0
    if item[0]<'Q':
        return 1
    return 2

for group, frame in df.groupby(fun):
    print('There are ' + str(len(frame)) + ' records in group ' + str(group) + ' for processing.')


In [ ]:
df.groupby('STNAME').agg({'CENSUS2010POP': np.average})

# More CSV datas

In [ ]:
# csv files
df = pd.read_csv('pokemon_data.csv')
#print(df.head(5)) #top 5 rows
# to see bottom
print(df.tail(3)) # bottom 3 rows
#df['HP']

In [ ]:
# excel files
import pandas as pd
import xlrd # for excel
df_xlsx = pd.read_excel('pokemon_data.xlsx')
print(df_xlsx.tail(3))

In [ ]:
# text files
# since the text file is created with tabs so we need to delete that
df = pd.read_csv('pokemon_data.txt', delimiter='\t')

print(df.head(5))

In [ ]:
# Read Headers
df.columns

In [ ]:
# to see the index content
df.Name

In [ ]:
# to see specific columns
print(df[['Name','HP','Speed']])# [[]] for multiple column AND  [] for single

In [ ]:
# Read Each Row
print(df.iloc[0:4])

In [ ]:
# Read a specific location (R,C)
print(df.iloc[2,1])

In [ ]:
for index, row in df.iterrows():
    print(index, row['Name'])

In [ ]:
# to look for specific type
df.loc[df['Type 1'] == "Grass"]

# Sorting datas

In [ ]:
df.sort_values(['Type 1', 'HP'], ascending=[1,0])

## Scales

In [ ]:
df = pd.DataFrame(['A+', 'A', 'A-', 'B+', 'B', 'B-', 'C+', 'C', 'C-', 'D+', 'D'],
                  index=['excellent', 'excellent', 'excellent', 'good', 'good', 'good', 'ok', 'ok', 'ok', 'poor', 'poor'])
df.rename(columns={0: 'Grades'}, inplace=True)
df

In [ ]:
df['Grades'].astype('category').head()

## Working in a time data frame

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
pd.Timestamp('9/1/2016 10:05AM')

In [ ]:
pd.Period('3/5/2016')

In [ ]:
pd.Period('5/2016')

In [ ]:
t1 = pd.Series(list('abc'), [pd.Timestamp('2016-09-01'), pd.Timestamp('2016-09-02'), pd.Timestamp('2016-09-03')])
t1

In [ ]:
d1 = ['2 June 2013', 'Aug 29, 2014', '2015-06-26', '7/12/16']
ts3 = pd.DataFrame(np.random.randint(10, 100, (4,2)), index=d1, columns=list('ab'))
ts3

In [ ]:
ts3.index = pd.to_datetime(ts3.index)
ts3

In [ ]:
pd.Timestamp('9/3/2016')-pd.Timestamp('9/1/2016')

In [ ]:
pd.Timestamp('9/2/2016 8:10AM') + pd.Timedelta('12D 3H')

In [ ]:
dates = pd.date_range('10-01-2020', periods=9, freq='2W-SUN')
dates

# Pandas Visualization

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib notebook

In [ ]:
# see the pre-defined styles provided.
plt.style.available

In [ ]:
# use the 'seaborn-colorblind' style
plt.style.use('seaborn-colorblind')

### DataFrame.plot

In [ ]:
np.random.seed(123)

df = pd.DataFrame({'A': np.random.randn(365).cumsum(0), 
                   'B': np.random.randn(365).cumsum(0) + 20,
                   'C': np.random.randn(365).cumsum(0) - 20}, 
                  index=pd.date_range('11-8-2020', periods=365))
df.head()

In [ ]:
df.plot(); # add a semi-colon to the end of the plotting call to suppress unwanted output

We can select which plot we want to use by passing it into the 'kind' parameter.

In [ ]:
df.plot('A','B', kind = 'scatter');

You can also choose the plot kind by using the `DataFrame.plot.kind` methods instead of providing the `kind` keyword argument.

`kind` :
- `'line'` : line plot (default)
- `'bar'` : vertical bar plot
- `'barh'` : horizontal bar plot
- `'hist'` : histogram
- `'box'` : boxplot
- `'kde'` : Kernel Density Estimation plot
- `'density'` : same as 'kde'
- `'area'` : area plot
- `'pie'` : pie plot
- `'scatter'` : scatter plot
- `'hexbin'` : hexbin plot

In [ ]:
# create a scatter plot of columns 'A' and 'C', with changing color (c) and size (s) based on column 'B'
df.plot.scatter('A', 'C', c='B', s=df['B'], colormap='viridis')

In [ ]:
ax = df.plot.scatter('A', 'C', c='B', s=df['B'], colormap='viridis')
ax.set_aspect('equal')

In [ ]:
df.plot.box();

In [ ]:
df.plot.hist(alpha=0.7);

[Kernel density estimation plots](https://en.wikipedia.org/wiki/Kernel_density_estimation) are useful for deriving a smooth continuous function from a given sample.

In [ ]:
df.plot.kde();

### pandas.tools.plotting

[Iris flower data set](https://en.wikipedia.org/wiki/Iris_flower_data_set)

In [ ]:
import pandas as pd
iris = pd.read_csv('iris.csv')
iris.head()

In [ ]:
pd.tools.plotting.scatter_matrix(iris);

# Seaborn

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib notebook

In [ ]:
np.random.seed(1234)

v1 = pd.Series(np.random.normal(0,10,1000), name='v1')
v2 = pd.Series(2*v1 + np.random.normal(60,15,1000), name='v2')

In [ ]:
plt.figure()
plt.hist(v1, alpha=0.7, bins=np.arange(-50,150,5), label='v1');
plt.hist(v2, alpha=0.7, bins=np.arange(-50,150,5), label='v2');
plt.legend();

In [ ]:
plt.figure()
plt.hist(v1, alpha=0.7, bins=np.arange(-50,150,5), label='v1');
plt.hist(v2, alpha=0.7, bins=np.arange(-50,150,5), label='v2');
plt.legend();

In [ ]:
plt.figure()
# we can pass keyword arguments for each individual component of the plot
sns.distplot(v3, hist_kws={'color': 'Teal'}, kde_kws={'color': 'Navy'});


In [ ]:
iris = pd.read_csv('iris.csv')
iris.head()

In [ ]:
sns.pairplot(iris, hue='Name', diag_kind='kde', size=2);

In [ ]:
plt.figure(figsize=(8,6))
plt.subplot(121)
sns.swarmplot('Name', 'PetalLength', data=iris);
plt.subplot(122)
sns.violinplot('Name', 'PetalLength', data=iris);